core/src/main/scala/kafka/server/DelayedOperation.scala

```diff
@@ -370,6 +383,20 @@ final class DelayedOperationPurgatory[T <: DelayedOperation](purgatoryName: Stri
 
       purged
     }
+
+    def forceComplete(): Int = {
```

Comment:
Actually, we can't hold the leader lock while checking the purgatory. Another way to do that is to remember the expected leader epoch in the delayed produce request. After the leader changes, the leader epoch will be advanced. If the delayed produce request errors out on non-matching leader epoch, then a tryComplete() call on the purgatory is guaranteed to complete all outstanding requests.